From: Bruno Haible <bruno@clisp.org>
Date: Fri, 14 Feb 2025 10:47:48 +0000 (+0100)
Subject: Fix a buffer overrun in the GB18030 converter (regression 2023-05-29).
X-Git-Url: https://git.savannah.gnu.org/gitweb/?p=libiconv.git;a=commitdiff_plain;h=d68ea07b28aa3c8d1959358b7da7e7f3ba148319;hp=bc17565f9a4caca27161609c526b776287a8270e

Fix a buffer overrun in the GB18030 converter (regression 2023-05-29).

Reported by Nick Wellnhofer <wellnhofer@aevum.de>.

* lib/gb18030_2005.h (gb18030_2005_wctomb): Return RET_TOOSMALL if there
is not enough room for 4 bytes.
---

diff --git a/ChangeLog b/ChangeLog
index 9fa38fc2da4a..df02b5fe4422 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,10 @@
+2025-02-14  Bruno Haible  <bruno@clisp.org>
+
+	Fix a buffer overrun in the GB18030 converter (regression 2023-05-29).
+	Reported by Nick Wellnhofer <wellnhofer@aevum.de>.
+	* lib/gb18030_2005.h (gb18030_2005_wctomb): Return RET_TOOSMALL if there
+	is not enough room for 4 bytes.
+
 2024-12-15  Bruno Haible  <bruno@clisp.org>
 
 	Update translations.
diff --git a/lib/gb18030_2005.h b/lib/gb18030_2005.h
index 4e7fca5..2da50d1 100644
--- a/lib/gb18030_2005.h
+++ b/lib/gb18030_2005.h
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 1999-2001, 2005, 2012, 2016, 2023 Free Software Foundation, Inc.
+ * Copyright (C) 1999-2001, 2005, 2012, 2016, 2023, 2025 Free Software Foundation, Inc.
  * This file is part of the GNU LIBICONV Library.
  *
  * The GNU LIBICONV Library is free software; you can redistribute it
@@ -361,11 +361,14 @@ gb18030_2005_wctomb (conv_t conv, unsigned char *r, ucs4_t wc, size_t n)
               r[1] = c & 0xff;
               return 2;
             } else {
-              r[0] = (c >> 24);
-              r[1] = (c >> 16) & 0xff;
-              r[2] = (c >> 8) & 0xff;
-              r[3] = c & 0xff;
-              return 4;
+              if (n >= 4) {
+                r[0] = (c >> 24);
+                r[1] = (c >> 16) & 0xff;
+                r[2] = (c >> 8) & 0xff;
+                r[3] = c & 0xff;
+                return 4;
+              } else
+                return RET_TOOSMALL;
             }
           }
         }
